dist/tools/kconfiglib/genconfig.py

```diff
@@ -65,6 +78,54 @@ def merge_configs(kconf, configs=[]):
         logging.debug(kconf.load_config(config, replace=False))
 
 
+def check_configs(kconf, check_app_sym=False):
+    """
+    Verifies that the generated configuration is valid.
+
+    A configuration is not valid when:
+        - An error symbol has been set because of an invalid condition.
+        - A module could not be set to the value defined by the user.
+        - A configuration parameter could not be set to value defined by the
+          user.
+    """
+    for sym in kconf.unique_defined_syms:
+        if is_error(sym) and sym.str_value:
+            logging.error("{}".format(sym.str_value))
+            return False
+
+        elif check_app_sym and sym.name == "APPLICATION" and sym.str_value == "n":
+            logging.error("The dependencies for the application are not"
+                          " met:\n{}".format(sym))
+            return False
```

Comment:
First impression on the interface here, this is what I get now when building an application that doesn't have the features:

```
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-The dependencies for the application are not met:
config APPLICATION
	bool
	default y
	depends on HAS_BACKUP_RAM && TEST_KCONFIG
make: Nothing to be done for 'clean'.
Building application "tests_periph_backup_ram" for "samr21-xpro" with MCU "samd21".

make: *** No rule to make target '/home/francisco/workspace/RIOT3/tests/periph_backup_ram/bin/samr21-xpro/generated/out.config', needed by '/home/francisco/workspace/RIOT3/tests/periph_backup_ram/bin/samr21-xpro/generated/autoconf.h'.  Stop.
make: Leaving directory '/home/francisco/workspace/RIOT3/tests/periph_backup_ram'
```

Some comments:

- IMO it should abort before the attempting to build and not because the out file is not there.
- would be nice to get colors with the error message just as it is now.
- its not clear which dependency is not met, here it is pretty easy to check, but it could be a bigger list.

Reply:
Similar comment for `CONFLICTING` configurations:

```
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-On SAM0, the RTC and RTT map to the same hardware peripheral.
Building application "tests_periph_rtc" for "samr21-xpro" with MCU "samd21".

make: *** No rule to make target '/home/francisco/workspace/RIOT3/tests/periph_rtc/bin/samr21-xpro/generated/out.config', needed by '/home/francisco/workspace/RIOT3/tests/periph_rtc/bin/samr21-xpro/generated/autoconf.h'.  Stop.
```

Would be nice to:

- have colored error messages
- for it to stop before attempting to compile

Reply:
Also I'm realizing that I get `=== [ATTENTION] Testing Kconfig dependency modelling ===` twice every time it compiles successfully.

```
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_periph_rtc" for "samr21-xpro" with MCU "samd21".

"make" -C /home/francisco/workspace/RIOT3/boards/samr21-xpro
"make" -C /home/francisco/workspace/RIOT3/core
"make" -C /home/francisco/workspace/RIOT3/cpu/samd21
```

Reply:
> Also I'm realizing that I get === [ATTENTION] Testing Kconfig dependency modelling === twice every time it compiles successfully.

yes, this is because Make restarts to load the configuration once it's produced.

Reply:
> for it to stop before attempting to compile
There is no compilation attempt, this is the step of configuration generation. I don't know what you mean by doing it before compiling here.

Reply:
Regarding the double message, see:
https://github.com/RIOT-OS/RIOT/blob/ad9e35c44562c2d5f7b891646508fd929ba4c487/makefiles/kconfig.mk#L46-L49